pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_database.go

```diff
@@ -0,0 +1,348 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package scbuildstmt
+
+import (
+	"strings"
+
+	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/security/username"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/catpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/decodeusername"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
+	"github.com/cockroachdb/cockroach/pkg/sql/privilege"
+	"github.com/cockroachdb/cockroach/pkg/sql/roleoption"
+	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scerrors"
+	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/catconstants"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlclustersettings"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlerrors"
+	"github.com/cockroachdb/cockroach/pkg/util/errorutil/unimplemented"
+	"github.com/cockroachdb/errors"
+)
+
+func CreateDatabase(b BuildCtx, n *tree.CreateDatabase) {
+	// TODO (xiang): Remove the fallback cases.
+	fallBackIfMultiRegion(b, n)
+
+	// 1. Run a bunch of pre-checks.
+	createDatabasePreChecks(b, n)
+	b.IncrementSchemaChangeCreateCounter("database")
+
+	// 2. Ensure database name is not already in use.
+	dbName := string(n.Name)
+	if databaseExists(b, n.Name) {
+		if n.IfNotExists {
+			return
+		}
+		panic(sqlerrors.NewDatabaseAlreadyExistsError(dbName))
+	}
+
+	// 3. Ensure current user can create a database with possibly a
+	// different user as the database's owner.
+	owner := getOwnerOfDB(b, n)
+	if err := checkCanAlterToOwner(b, n, owner); err != nil {
+		panic(err)
+	}
+
+	// 4. Construct and add all relevant elements, including the database,
+	// its public schema, and others.
+	dbElem, _ := addCreateDatabaseElements(b, dbName, owner)
+	b.LogEventForExistingTarget(dbElem)
+}
+
+func addCreateDatabaseElements(
+	b BuildCtx, dbName string, dbOwner username.SQLUsername,
+) (*scpb.Database, *scpb.Schema) {
+	databaseID := b.GenerateUniqueDescID()
+	publicSchemaID := b.GenerateUniqueDescID()
+
+	// Add database related elements.
+	dbElem := &scpb.Database{DatabaseID: databaseID}
+	b.Add(dbElem)
+	b.Add(&scpb.Namespace{
+		DatabaseID:   0,
+		SchemaID:     0,
+		DescriptorID: databaseID,
+		Name:         dbName,
+	})
+	b.Add(&scpb.DatabaseData{DatabaseID: databaseID})
+	b.Add(&scpb.DatabaseRoleSetting{
+		DatabaseID: databaseID,
+		RoleName:   scpb.PlaceHolderRoleName,
+	})
```

Comment:
Adding those two elements here does not actually do anything, but needed here to correctly handle cases like
```
create database db;
drop database db;
```
where the `drop database` would decompose and generate `DatabaseData` and `DatabaseRoleSetting` elements.